# `pandas` Part 7: Combining Datasets with `concat()`

# Learning Objectives
## By the end of this tutorial you will be able to:
1. Combine DataFrames and/or Series with `concat()`
2. Understand a multi-index
3. Reset an index with `reset_index()`
4. Perform descriptive analytics on a combined DataFrame

## Files Needed for this lesson:
>- `CAvideos.csv`
>- `GBvideos.csv`
>- Download this csv files from Canvas prior to the lesson
>- C:\\Users\\mimc2537\\OneDrive - UCB-O365\\python\\pandas\\

## The general steps to working with pandas:
1. import pandas as pd
2. Create or load data into a pandas DataFrame or Series
3. Reading data with `pd.read_`
>- Excel files: `pd.read_excel('fileName.xlsx')`
>- Csv files: `pd.read_csv('fileName.csv')`
>- Note: if the file you want to read into your notebook is not in the same folder you can do one of two things:
>>- Move the file you want to read into the same folder/directory as the notebook
>>- Type out the full path into the read function
4. After steps 1-3 you will want to check out your DataFrame
>- Use `shape` to see how many records and columns are in your DataFrame
>- Use `head()` to show the first 5-10 records in your DataFrame

# Introduction Notes on Combining Data Using `pandas`
1. Being able to combine data from multiple sources is a critical skill for analytics professionals
2. We will learn the `pandas` way of combining data but there are similarities here to SQL
3. Why combine data with `pandas` if you can do the same thing in SQL?
>- The answer to this depends on the project
>- Some projects may be completed more efficiently all with `pandas` so you wouldn't necessarily need SQL
>- For some projects incorporating SQL into our python code makes sense
>- In a an analytics job, you will likely use both python and SQL but this is a python class so we will learn the python way! 

# Initial set-up steps
1. import modules and check working directory
2. Read data in
3. Check the data

In [32]:
import pandas as pd; import os

path = os.getcwd()
path

'C:\\Users\\anton\\Documents\\CU-Python\\week_13'

# Step 2 Read Data Into a DataFrame with `read_csv()`
>- file names: 
>>- `CAvideos.csv`
>>- `GBvideos.csv`

In [33]:
ca = pd.read_csv(path+'\\CAvideos.csv')
gb = pd.read_csv(path+'\\GBvideos.csv')

### Check how many rows and columns are in our DataFrames

In [34]:
print('The rows and cols in this DataFrame lesson')
print(f'\tCanadian YouTube df: {ca.shape}, \n\tUK YouTube df: {gb.shape}')

#double check that they have the same number of cols and will need to make sure they align to use concat

The rows and cols in this DataFrame lesson
	Canadian YouTube df: (40881, 16), 
	UK YouTube df: (38916, 16)


### Check a couple of rows of data in one of the new DataFrames

In [35]:
ca.head(2)

video_id trending_date                                       title  \
0  n1WpP7iowLc      17.14.11  Eminem - Walk On Water (Audio) ft. Beyoncé   
1  0dBIkQ4Mz1M      17.14.11               PLUSH - Bad Unboxing Fan Mail   

  channel_title  category_id              publish_time  \
0    EminemVEVO           10  2017-11-10T17:00:03.000Z   
1     iDubbbzTV           23  2017-11-13T17:00:00.000Z   

                                                tags     views   likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579  787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651  127794   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   

                                         description  
0  Eminem's new track Walk on Water ft. Beyoncé i...  
1  STill got a lot of packages. Probably will las...

In [36]:
gb.head(2)

video_id trending_date                                          title  \
0  Jw1Y-zhQURU      17.14.11  John Lewis Christmas Ad 2017 - #MozTheMonster   
1  3s1rvMFUweQ      17.14.11      Taylor Swift: …Ready for It? (Live) - SNL   

         channel_title  category_id              publish_time  \
0           John Lewis           26  2017-11-10T07:38:29.000Z   
1  Saturday Night Live           24  2017-11-12T06:24:44.000Z   

                                                tags    views  likes  \
0  christmas|"john lewis christmas"|"john lewis"|...  7224515  55681   
1  SNL|"Saturday Night Live"|"SNL Season 43"|"Epi...  1053632  25561   

   dislikes  comment_count                                  thumbnail_link  \
0     10247           9479  https://i.ytimg.com/vi/Jw1Y-zhQURU/default.jpg   
1      2294           2757  https://i.ytimg.com/vi/3s1rvMFUweQ/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   

                                         description  
0  Click here to continue the story and make your...  
1  Musical guest Taylor Swift performs …Ready for...

## Check the datatypes

In [37]:
ca.dtypes

video_id                  object
trending_date             object
title                     object
channel_title             object
category_id                int64
publish_time              object
tags                      object
views                      int64
likes                      int64
dislikes                   int64
comment_count              int64
thumbnail_link            object
comments_disabled           bool
ratings_disabled            bool
video_error_or_removed      bool
description               object
dtype: object

# Combining DataFrames
>- The three common ways to combine datastest in pandas is with `concat()`, `join()`, and `merge()`
>- `concat()` will take two DataFrames or Series and append them together
>>- This is basically taking DataFrames and stacking their data on top of each other into one DataFrame
>>- For `concat()` you need the columns/fields in both DataFrames to the be the same
>- `join()` "links" DataFrames together based on a common field/column between the two
>- `merge()` also links DataFrames together based on common field/columns but with different syntax.
>>- We will cover the most basic join in this class
>>- A more in depth study of joins is provided in BAIM 4205 - Business Data Management (SQL)
>>- Pandas join reference for further study: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html


# Using the YouTube DataFrames to practice combining data with pandas
>- The YouTube datasets store data on various YouTube trending statistics
>- Our example datasets show several months of data and daily trending YouTube videos.

>- For more information and for other YouTube datasets see the following link:
>>- https://www.kaggle.com/datasnaek/youtube-new

### First, creating a new DataFrame that appends the Canadian and British YouTube DataFrames

In [40]:
ca_gb = pd.concat([ca,gb], 
                  keys=['can','UK'], #keys need to line up with dataframes 
                  names=['country', 'rowid_country']) #lable the index column that was defined in the keys function

#### Some notes on the previous code
>- Line 1: We define a new DataFrame named `CanUK` which is defined as the concatenation,`concat()`, of two datasets
>>- Dataset 1 = canadian_youtube
>>- Dataset 2 = uk_youtube
>>- The `concat()` function takes the two (or more if applicable) DataFrames and "stacks" them on top of each
>- Line 2: We use `keys` option to define a multi-index (aka hierarchical index)
>>- Because our datasets represent YouTube videos from different countries we pass the abbreviated names of those countries as a list to `keys`
>>- Enter the keys names in order they appear in line 1 (e.g., 'can' first, 'uk' second)
>- Line 3: We use the `names` option to label our index columns from line 2
>>- Without the `names` option we would not have anything above our index columns

### Check the index for any dataframe using `DataFrame.index`
>- Note how `concat()` uses the rowid's for each country's dataset versus continuing the count

In [44]:
ca_gb.index #which rows are can, which rows are UK - shows how the dataframes are stacked 

MultiIndex([('can',     0),
            ('can',     1),
            ('can',     2),
            ('can',     3),
            ('can',     4),
            ('can',     5),
            ('can',     6),
            ('can',     7),
            ('can',     8),
            ('can',     9),
            ...
            ( 'UK', 38906),
            ( 'UK', 38907),
            ( 'UK', 38908),
            ( 'UK', 38909),
            ( 'UK', 38910),
            ( 'UK', 38911),
            ( 'UK', 38912),
            ( 'UK', 38913),
            ( 'UK', 38914),
            ( 'UK', 38915)],
           names=['country', 'rowid_country'], length=79797)

### Take a look at our new DataFrame

In [45]:
ca_gb.head(2) #see the country and rowid_country

video_id trending_date  \
country rowid_country                              
can     0              n1WpP7iowLc      17.14.11   
        1              0dBIkQ4Mz1M      17.14.11   

                                                            title  \
country rowid_country                                               
can     0              Eminem - Walk On Water (Audio) ft. Beyoncé   
        1                           PLUSH - Bad Unboxing Fan Mail   

                      channel_title  category_id              publish_time  \
country rowid_country                                                        
can     0                EminemVEVO           10  2017-11-10T17:00:03.000Z   
        1                 iDubbbzTV           23  2017-11-13T17:00:00.000Z   

                                                                    tags  \
country rowid_country                                                      
can     0              Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...   
        1              plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   

                          views   likes  dislikes  comment_count  \
country rowid_country                                              
can     0              17158579  787425     43420         125882   
        1               1014651  127794      1688          13030   

                                                       thumbnail_link  \
country rowid_country                                                   
can     0              https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
        1              https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   

                       comments_disabled  ratings_disabled  \
country rowid_country                                        
can     0                          False             False   
        1                          False             False   

                       video_error_or_removed  \
country rowid_country                           
can     0                               False   
        1                               False   

                                                             description  
country rowid_country                                                     
can     0              Eminem's new track Walk on Water ft. Beyoncé i...  
        1              STill got a lot of packages. Probably will las...

In [48]:
ca_gb.tail(2) #look at the last 2 entries

video_id trending_date  \
country rowid_country                              
UK      38914          -DRsfNObKIQ      18.14.06   
        38915          4YFo4bdMO8Q      18.14.06   

                                                                   title  \
country rowid_country                                                      
UK      38914          Eleni Foureira - Fuego - Cyprus - LIVE - First...   
        38915          KYLE - Ikuyo feat.  2 Chainz & Sophia Black [A...   

                                 channel_title  category_id  \
country rowid_country                                         
UK      38914          Eurovision Song Contest           24   
        38915                   SuperDuperKyle           10   

                                   publish_time  \
country rowid_country                             
UK      38914          2018-05-08T20:32:32.000Z   
        38915          2018-05-11T04:06:35.000Z   

                                                                    tags  \
country rowid_country                                                      
UK      38914          Eurovision Song Contest|"2018"|"Lisbon"|"Cypru...   
        38915          Kyle|"SuperDuperKyle"|"Ikuyo"|"2 Chainz"|"Soph...   

                          views   likes  dislikes  comment_count  \
country rowid_country                                              
UK      38914          14317515  151870     45875          26766   
        38915            607552   18271       274           1423   

                                                       thumbnail_link  \
country rowid_country                                                   
UK      38914          https://i.ytimg.com/vi/-DRsfNObKIQ/default.jpg   
        38915          https://i.ytimg.com/vi/4YFo4bdMO8Q/default.jpg   

                       comments_disabled  ratings_disabled  \
country rowid_country                                        
UK      38914                      False             False   
        38915                      False             False   

                       video_error_or_removed  \
country rowid_country                           
UK      38914                           False   
        38915                           False   

                                                             description  
country rowid_country                                                     
UK      38914          Eleni Foureira represented Cyprus at the first...  
        38915          Debut album 'Light of Mine' out now: http://ky...

In [49]:
ca_gb.shape

(79797, 16)

#### Did using `concat()` work to append the two DataFrames together? 
>- Check the shape of your new DataFrame
>- Compare the number of records to each one individually
>>- canadian_youtube = 40881 records
>>- uk_youtube = 38916 records
>>- 40881 + 38916 = 79797 total records

#### `reset_index`:
##### Note: You can reset a an index with `reset_index` 
>- This can be useful for some situations
>- For a multi-index you can pass the `level` option and specify what index you want to reset
>- Note: To make the change to our current DataFrame we would need to use the option, `inplace=True`

In [51]:
ca_gb.reset_index(level='country').head(2) #pass inplace=True to make permanent. Country is now a normal column instead of a multi-index

country     video_id trending_date  \
rowid_country                                      
0                 can  n1WpP7iowLc      17.14.11   
1                 can  0dBIkQ4Mz1M      17.14.11   

                                                    title channel_title  \
rowid_country                                                             
0              Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                           PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   

               category_id              publish_time  \
rowid_country                                          
0                       10  2017-11-10T17:00:03.000Z   
1                       23  2017-11-13T17:00:00.000Z   

                                                            tags     views  \
rowid_country                                                                
0              Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579   
1              plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651   

                likes  dislikes  comment_count  \
rowid_country                                    
0              787425     43420         125882   
1              127794      1688          13030   

                                               thumbnail_link  \
rowid_country                                                   
0              https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1              https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   

               comments_disabled  ratings_disabled  video_error_or_removed  \
rowid_country                                                                
0                          False             False                   False   
1                          False             False                   False   

                                                     description  
rowid_country                                                     
0              Eminem's new track Walk on Water ft. Beyoncé i...  
1              STill got a lot of packages. Probably will las...

In [53]:
ca_gb.head(3) #show that the change was not permanent (country is multi-index again)

video_id trending_date  \
country rowid_country                              
can     0              n1WpP7iowLc      17.14.11   
        1              0dBIkQ4Mz1M      17.14.11   
        2              5qpjK5DgCt4      17.14.11   

                                                                   title  \
country rowid_country                                                      
can     0                     Eminem - Walk On Water (Audio) ft. Beyoncé   
        1                                  PLUSH - Bad Unboxing Fan Mail   
        2              Racist Superman | Rudy Mancuso, King Bach & Le...   

                      channel_title  category_id              publish_time  \
country rowid_country                                                        
can     0                EminemVEVO           10  2017-11-10T17:00:03.000Z   
        1                 iDubbbzTV           23  2017-11-13T17:00:00.000Z   
        2              Rudy Mancuso           23  2017-11-12T19:05:24.000Z   

                                                                    tags  \
country rowid_country                                                      
can     0              Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...   
        1              plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   
        2              racist superman|"rudy"|"mancuso"|"king"|"bach"...   

                          views   likes  dislikes  comment_count  \
country rowid_country                                              
can     0              17158579  787425     43420         125882   
        1               1014651  127794      1688          13030   
        2               3191434  146035      5339           8181   

                                                       thumbnail_link  \
country rowid_country                                                   
can     0              https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
        1              https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
        2              https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   

                       comments_disabled  ratings_disabled  \
country rowid_country                                        
can     0                          False             False   
        1                          False             False   
        2                          False             False   

                       video_error_or_removed  \
country rowid_country                           
can     0                               False   
        1                               False   
        2                               False   

                                                             description  
country rowid_country                                                     
can     0              Eminem's new track Walk on Water ft. Beyoncé i...  
        1              STill got a lot of packages. Probably will las...  
        2              WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...

# Now some descriptive analytics

### What channels have the most trending videos?

In [57]:
ca_gb.groupby(['channel_title']).title.count().sort_values(ascending= False) #doesn't need to be title I dont think

channel_title
The Late Show with Stephen Colbert        361
TheEllenShow                              357
Jimmy Kimmel Live                         351
The Tonight Show Starring Jimmy Fallon    345
Late Night with Seth Meyers               342
                                         ... 
Mansoor                                     1
Manorama News                               1
Manorama Music Songs                        1
Mann                                        1
MingJunn Goh                                1
Name: title, Length: 5956, dtype: int64

### What are the quantitative descriptive statistics for TheEllenShow?

In [59]:
ca_gb.where(ca_gb.channel_title == 'TheEllenShow').describe() #only applies describe on rows where the channel title is TheEllenShow

category_id         views          likes      dislikes  comment_count  \
count        357.0  3.570000e+02     357.000000    357.000000     357.000000   
mean          24.0  1.779820e+06   46257.770308   1328.537815    2179.221289   
std            0.0  2.008212e+06   62550.222512   3198.710961    3614.375611   
min           24.0  5.631300e+04    1187.000000     18.000000       0.000000   
25%           24.0  5.360520e+05   14152.000000    243.000000     318.000000   
50%           24.0  1.238747e+06   28108.000000    598.000000     940.000000   
75%           24.0  2.022094e+06   59950.000000   1309.000000    2858.000000   
max           24.0  1.359265e+07  499673.000000  26397.000000   28381.000000   

       comments_disabled  ratings_disabled  video_error_or_removed  
count         357.000000             357.0                   357.0  
mean            0.170868               0.0                     0.0  
std             0.376922               0.0                     0.0  
min             0.000000               0.0                     0.0  
25%             0.000000               0.0                     0.0  
50%             0.000000               0.0                     0.0  
75%             0.000000               0.0                     0.0  
max             1.000000               0.0                     0.0

##### Alternatively, you can use `loc[]` to peform the filtering operation
>- The use of `where()` or `loc[]` depends on the question/purpose or sometimes just personal preference

In [61]:
ca_gb.loc[ca_gb.channel_title=='TheEllenShow'].describe() #does not return info on non numeric data

category_id         views          likes      dislikes  comment_count
count        357.0  3.570000e+02     357.000000    357.000000     357.000000
mean          24.0  1.779820e+06   46257.770308   1328.537815    2179.221289
std            0.0  2.008212e+06   62550.222512   3198.710961    3614.375611
min           24.0  5.631300e+04    1187.000000     18.000000       0.000000
25%           24.0  5.360520e+05   14152.000000    243.000000     318.000000
50%           24.0  1.238747e+06   28108.000000    598.000000     940.000000
75%           24.0  2.022094e+06   59950.000000   1309.000000    2858.000000
max           24.0  1.359265e+07  499673.000000  26397.000000   28381.000000

### What were the total YouTube videos, total views, likes and dislikes for TheEllenShow?
>- Using the agg() function to calculate specific aggregations on different columns

In [63]:
ca_gb.where(ca_gb.channel_title=='TheEllenShow').agg({'title':['count'],
                                                      'views': ['sum'],
                                                      'likes': ['sum'],
                                                      'dislikes': ['sum']
                                                     }) #can pass more aggregations in the list (count, mean, etc...)

title        views       likes  dislikes
count  357.0          NaN         NaN       NaN
sum      NaN  635395636.0  16514024.0  474288.0

## What are the totals for TheEllenShow's top 5 most viewed videos?
>- Only include the title names as part of the output (not channel or any other categorical fields)
>- Include total views, likes, dislikes, and comment count in the output

### Technique for answering Q's
>- Start by defining the problem and determinging what your code will need to do --- 
 
        1. "Need to filter only TheEllenShow
        2. need to only have the 5 columns
        3. need to determine the most viewed videos
        5. need to show only 5

In [72]:
ca_gb.loc[ca_gb.channel_title=='TheEllenShow', 
         ['title', 'views', 'likes', 'dislikes', 'comment_count']].\
    groupby(['title']).sum().\
    sort_values(by ='views', ascending= False).\
    head(5) ##Ellen show only, #the desired columns only, #group by title because they are asking for top videos, #sort by the number of views becasue 'top 5 most viewed', top 5 only

views    likes  \
title                                                                   
Jennifer Lawrence Explains Her Drunk Alter Ego ...  71621685  1302161   
Kid Yodeler Mason Ramsey Performs                   68553312  2697959   
Kim Kardashian Lets Gender of Third Child Slip      52985123   776177   
Billionaire Bill Gates Guesses Grocery Store Pr...  37076353   542804   
Laurel or Yanny?                                    28695358   613487   

                                                    dislikes  comment_count  
title                                                                        
Jennifer Lawrence Explains Her Drunk Alter Ego ...     35255          49762  
Kid Yodeler Mason Ramsey Performs                     144604              0  
Kim Kardashian Lets Gender of Third Child Slip         26509          50940  
Billionaire Bill Gates Guesses Grocery Store Pr...     29036          42720  
Laurel or Yanny?                                       18568         100731

# Some Notes on the Previous Example
>- Our pandas code in the previous example is similar to SQL in the following ways
    1. `loc[CanUk.channel_title == 'TheEllenShow',` is SQL equivalent to `WHERE channel_title = 'TheEllenShow'`
    2. `['title','views','likes','dislikes','comment_count']` is SQL equivalent to:
        `SELECT title, sum(views),sum(likes),sum(dislikes),sum(comment_count)`
    3. `groupby(['title`]) is SQL equivalent to GROUP BY title
    4. Now in pandas we enter the aggregation after the `groupby()`, in this example `sum()`
      >>- In SQL we write the aggregation in the SELECT statement
      
## In future lessons we will continue to learn how pandas and SQL relate